backends-velox/src/main/scala/org/apache/spark/sql/execution/BroadcastUtils.scala

```diff
@@ -172,19 +172,21 @@ object BroadcastUtils {
     if (filtered.isEmpty) {
       return ColumnarBatchSerializeResult.EMPTY
     }
-    val handleArray =
-      filtered.map(b => ColumnarBatches.getNativeHandle(BackendsApiManager.getBackendName, b))
-    val serializeResult =
-      try {
-        ColumnarBatchSerializerJniWrapper
-          .create(
-            Runtimes
+    var rowNums = 0
+    val values = filtered.map(
+      b => {
+        val handle = ColumnarBatches.getNativeHandle(BackendsApiManager.getBackendName, b)
+        rowNums += b.numRows()
+        try {
+          ColumnarBatchSerializerJniWrapper
+            .create(Runtimes
               .contextInstance(BackendsApiManager.getBackendName, "BroadcastUtils#serializeStream"))
-          .serialize(handleArray)
-      } finally {
-        filtered.foreach(ColumnarBatches.release)
-      }
-    serializeResult
+            .serialize(handle)
+        } finally {
+          ColumnarBatches.release(b)
```

Comment:
So release 'b' every loop can reduce memory consumption?

Reply:
Thanks for your explains.

Reply:
@wForget 

Did you observe a decrement of peak off-heap usage after this change? As shown in code, `filtered` is still of the type `Array[ColumnarBatch]` which may cause high usage from the beginning of the loop.

Reply:
> Did you observe a decrement of peak off-heap usage after this change? 

After this change, job no longer has offheap OOM. Is this enough to indicate that offheap peek memory has decreased?

> As shown in code, filtered is still of the type Array[ColumnarBatch] which may cause high usage from the beginning of the loop.

Makes sense, this is indeed an improvement, I will try to convert it to an iterator